py3status/modules/weather_owm.py

```diff
@@ -0,0 +1,750 @@
+
+# Paths of information to extract from JSON
+IP_LAT = '//lat'
+IP_LNG = '//lon'
+OWM_CLOUD_COVER = '//clouds/all'
+OWM_DESC = '//weather:0/main'
+OWM_DESC_LONG = '//weather:0/description'
+OWM_HUMIDITY = '//main/humidity'
+OWM_PRESSURE = '//main'
+OWM_RAIN = '//rain/3h'
+OWM_SNOW = '//snow/3h'
+OWM_SUNRISE = '//sys/sunrise'
+OWM_SUNSET = '//sys/sunset'
+OWM_TEMP = '//main'
+OWM_WEATHER_ICON = '//weather:0/id'
+OWM_WIND = '//wind'
+
+# Units constants
+RAIN_UNITS = set(['cm', 'ft', 'in', 'mm', 'm', 'yd'])
+SNOW_UNITS = RAIN_UNITS
+TEMP_UNITS = set(['c', 'f', 'k'])
+WIND_UNITS = set(['fsec', 'msec', 'mph', 'kmh'])
+
+# Conversion factors
+FT_FROM_METER = 3.28084
+IN_FROM_MM = 0.0393701
+KMH_FROM_MSEC = 0.277778
+MPH_FROM_MSEC = 2.23694
+
+# Thresholds defaults
+THRESHOLDS = [
+    (-100, '#0FF'),
+    (0, '#00F'),
+    (50, '#0F0'),
+    (150, '#FF0')
+]
+
+
+class OWMException(Exception):
+    pass
+
+
+class Py3status:
+
+    api_key = None
+    cache_timeout = 600
+    forecast_days = 3
+    forecast_include_today = False
+    forecast_text_separator = ' '
+    format = '{city}: {icon} {temperature} {main} {forecast}'
+    format_clouds = '{icon} {coverage}%'
+    format_forecast = '{icon}'
+    format_humidity = '{icon} {humidity}%'
+    format_pressure = '{icon} {pressure} hPa'
+    format_rain = '{icon} {amount:.0f} {unit}'
+    format_snow = '{icon} {amount:.0f} {unit}'
+    format_sunrise = '{icon} %-I:%M %p'
+    format_sunset = '{icon} %-I:%M %p'
+    format_temperature = u'{icon} {current:.0f}°{unit}'
+    format_wind = '{icon} {speed:.0f} {unit}'
+    icon_atmosphere = u'🌫'
+    icon_cloud = u'☁'
+    icon_extreme = u'⚠'
+    icon_humidity = u'●'
+    icon_pressure = u'◌'
+    icon_rain = u'🌧'
+    icon_snow = u'❄'
+    icon_sun = u'☼'
+    icon_temperature = u'○'
+    icon_thunderstorm = u'⛈'
+    icon_wind = u'☴'
+    icons = None
+    lang = 'en'
+    location = None
+    rain_unit = 'in'
+    request_timeout = 10
+    snow_unit = 'in'
+    temp_unit = 'f'
+    thresholds = THRESHOLDS
+    wind_unit = 'mph'
+
+    def _get_icons(self):
+        if self.icons is None:
+            self.icons = {}
+
+        # Defaults for weather ranges
+        defaults = {
+            'i200_i299': self.icon_thunderstorm,
+            'i300_i399': self.icon_rain,
+            'i500_i599': self.icon_rain,
+            'i600_i699': self.icon_snow,
+            'i700_i799': self.icon_atmosphere,
+            'i800': self.icon_sun,
+            'i801_i809': self.icon_cloud,
+            'i900_i909': self.icon_extreme,
+            'i950_i959': self.icon_wind,
+            'i960_i999': self.icon_extreme,
+        }
+
+        # Default mappings for other icons
+        others = {
+            'clouds': 802,
+            'rain': 501,
+            'snow': 601,
+            'wind': 954,
+            'humidity': self.icon_humidity,
+            'pressure': self.icon_pressure,
+            'temperature': self.icon_temperature,
```

Comment:
This works if the format have `{icon}` so maybe we need to keep them all supplied. Forget what I recently said about statically putting icons in various `format_xxx` if we were to to do this and/or make `icons = {}` fully workable. Example config.
```
        icons = {
            clouds:'CLOUD',
            humidity:'HUMIDITY',
            pressure:'PRESSURE',
            rain:'RAIN',
            snow:'SNOW',
            wind:'WIND',
            temperature: 'TEMPERATURE',
            sunset: 'SUNSET',
            sunrise: 'SUNRISE',
        }
```
Seeing how that's possible to customize icons using `icons = {xxx:x}` instead of outside `icon_xxx`... `icon_xxx` parts is now somewhat redundant. 

I don't know if @tobes / @ultrabug would like this idea. They shot down `button = {pause:1, stop:3, next:4, prev:5}` in new modules PRs. Your `icons = {xxx:x}` likely was made to work with range, but can simply work with standalone icons too (like what your desc said)... meaning we could live without all those `icon_xxx` config.

```
        Also, you can specify the icons for the various formatting sections
        below. For example, to specify the icon for 'format_pressure', use
        'pressure'. A few formatting sections will take from the defaults or be
        dynamic, including (clouds, rain, wind, sunrise, sunset).
        These may be specified regardless.
```